helix-term/src/ui/fuzzy_match.rs

```diff
@@ -0,0 +1,74 @@
+use fuzzy_matcher::skim::SkimMatcherV2 as Matcher;
+use fuzzy_matcher::FuzzyMatcher;
+
+#[cfg(test)]
+mod test;
+
+pub struct FuzzyQuery {
+    queries: Vec<String>,
+}
+
+impl FuzzyQuery {
+    pub fn new(query: &str) -> FuzzyQuery {
+        let mut saw_backslash = false;
+        let queries = query
+            .split(|c| {
+                saw_backslash = match c {
+                    ' ' if !saw_backslash => return true,
+                    '\\' => true,
+                    _ => false,
+                };
+                false
+            })
+            .filter_map(|query| {
+                if query.is_empty() {
+                    None
+                } else {
+                    Some(query.replace("\\ ", " "))
+                }
+            })
+            .collect();
+        FuzzyQuery { queries }
+    }
```

Comment:
```suggestion
    pub fn new(query: &str) -> FuzzyQuery {
        const SIGNISPACE_PLACEHOLDER: &str = "$$$$";
        let queries = query.replace("\\ ", SIGNISPACE_PLACEHOLDER)
            .split_whitespace()
            .map(|query| query.replace(SIGNISPACE_PLACEHOLDER, " "))
            .collect();
        FuzzyQuery { queries }
    }
```

Reply:
That looks hacky too me. It's possible that someone would want to search for `$$$$`.

Reply:
That's why I made it a const for easy replacement. I agree that `$$$$` is a bad choice, but maybe every choice is a bad one - and an FSM is better.

I did find the original code easier to read if the check against `\\` is first like:
```rust
                saw_backslash = match c {
                    '\\' => true,
                    ' ' if !saw_backslash => return true,
                    _ => false,
                };```
It kind of tells the reader what's going on in order of observed tokens.

Reply:
I marked the suggestion as resolved, I agree that it was a bit too hacky.

Reply:
I have written a ton of parsers/lexers recently (my job is working on a compiler) so this code was very natural to me but if you find it more readable I can definitely change the order of the match.

Your suggestion is actually somewhat similar to what`skim` uses internally actually.
However they use the null terminator '\0' as a placeholder.
The reasons they do it that way is because they use the `regex` crate instead of rolling their own implementation and regex can not support this kind of escape sequence.
They use regex because they actually support many more patterns in their matcher which I did not include in this PR because they are not the same across various implementations and I wanted to keep the PR uncontroversial.

In my experience working with compilers using a custom lexer often ends up being simpler (regex has a nasty habit of growing into monstrosities) and in many cases also more performant.

Also using `split_whitespace` here is not ideal because it matches any unicode white space (for example newlines) which is a behavior other tools/editors do not share and therefore do not want to introduce here.

Reply:
I think my confusion can be explained by the late hours and this was the first time I saw an if condition as part of the match case. 

What is the intended result of input `one two\ three four` ? I would expect it to break into `["one", "two three", "four"]` but I think it will be `["one", "two three four"]` ?

Reply:
No this works just fine, I tested in on the [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=ed75f071aa9ef2de64a2e3eb84a14a9d)

I will add a comment explaining the parser a bit more and a few more testcases but this is a very basic FSM so I am extremely confident it works as intended.

Reply:
seems like the early hours got me too, haha - the `_` fallback will of course toggle it back. Thanks for putting up with me.

BTW; no need to add superfluous comments on my account. Now that I look at it again with the `if` case match part of my vocabulary for Rust it is concise and readable.